sound/soc/sof/sof-audio.c

```diff
@@ -128,20 +135,27 @@ int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 		break;
 	}
 
+	/* continue to disable core even if IPC fails */
 	ret = sof_ipc_tx_message(sdev->ipc, ipc_free.hdr.cmd, &ipc_free, sizeof(ipc_free),
 				 &reply, sizeof(reply));
-	if (ret < 0) {
+	if (ret < 0)
 		dev_err(sdev->dev, "error: failed to free widget %s\n", swidget->widget->name);
-		swidget->use_count++;
-		return ret;
-	}
+
+	/*
+	 * disable widget core. continue to route setup status and complete flag
+	 * even if this fails
+	 */
+	if (snd_sof_dsp_core_put(sdev, core) < 0)
+		dev_err(sdev->dev, "error: failed to disable target core: %d for widget %s\n",
+			core, swidget->widget->name);
+
 
 	/* reset route setup status for all routes that contain this widget */
 	sof_reset_route_setup_status(sdev, swidget);
 	swidget->complete = 0;
 	dev_dbg(sdev->dev, "widget %s freed\n", swidget->widget->name);
 
-	return 0;
+	return ret;
```

Comment:
you can return a negative error code if the IPC failed but in fact you still performed the rest of the clean up. A failing IPC used to return an error from this before too, so you aren't changing that. There are many places in hda.c and hda-dai.c where this error code is propagated to the caller, hopefully they're all correct and are taking into account that this can be a "soft" failure.

Reply:
yes, the thing is the core ref_count needs to decremented even if the widget free fails. I think this is correct to do a core_put even if widget free failed. Otherwise, the core ref_count's will be messed up and will prevent D3